### ISSUE TITLE
parallel replication

### ISSUE THREAD

**Issue by problame** [zrepl/zrepl#140]
Someone wished for that feature at OpenZFS dev summit 2018.
Usecase was: many small home directory FSs, where HW wasn't the bottleneck.

#139 introduces stepQueue in the replication/driver package.
In that PR, it is used sequentially, i.e., only one FS is allowed to make progress at a time.
However, wrt replication/driver, this number could just as well be cranked up to parallel operation.

## What's missing?

The receiving side must be able to handle concurrent access to ZFS. Things to consider:

* placeholder creation must be serialized (concurrent placeholder creation will fail)
* no concurrent recv to a filesystem
  * not a real problem on the happy path, but we need to check for concurrent access anyways in case a connection goes stale and the client reconnects
    * reconnecting client should be able to preempt (cancel ctx) previous sessions
    * concurrent recvs should fail if it's the same session (bad client behavior assumed)
      * need TryLock or similiar
      * use appropriate gRPC status code?
* how does zfs handle zfs list with a concurrent zfs recv?
  * need to synchronize across different sink jobs?

Further, we need some knob to control the degree of concurrency allowed to stepQueue.

**Comment by problame** [comment]
Update:
```
diff --git a/replication/driver/replication_driver.go b/replication/driver/replication_driver.go
index 0c1cd8f..db26544 100644
--- a/replication/driver/replication_driver.go
+++ b/replication/driver/replication_driver.go
@@ -355,7 +355,7 @@ func (a *attempt) do(ctx context.Context, prev *attempt) {
        // invariant: prevs contains an entry for each unambigious correspondence
 
        stepQueue := newStepQueue()
-       defer stepQueue.Start(1)() // TODO parallel replication
+       defer stepQueue.Start(int(envconst.Int64("ZREPL_EXPERIMENTAL_REPLICATION_STEP_QUEUE_CONCURRENCY", 1)))() // TODO parallel replication
        var fssesDone sync.WaitGroup
        for _, f := range a.fss {
                fssesDone.Add(1)
```
Works flawlessly with a value of 20 (only tested if all target filesystems are already created, this is where i'd expect race conditions)

**Comment by problame** [comment]
Idea: Auto-tune parallelism for throughput?

Observations: TCP rounttrips + zfs send response time (time from starting zfs send to first byte received) can be significant
Additionally: `zfs send -n ` (pre redacted send&recv changes ?) consumes significant CPU time

Resources to monitor / optimize:
* minimize time spent waiting for rpcs (globally, i.e., count only while all senders are blocked on Send / recv request)
* maximize throughput (bytes / second) up to user-defined limit
* cpu utilization
  * relative thresholds are not useful, because zrepl could easily be allowed to monopolize cpu while idle, but have lower priority if anything else happens on the system
  * but relying on nice isn't great either, we would rather like to know that the system scheduler would throttle us in favor of another process, and we would reduce parallelism in response.
    * is there an interface for this?

**Comment by InsanePrawn** [comment]
I think the sooner we make this properly official, docs and config knobs and all, the sooner we find all those pesky race conditions in zrepl and zfs :)

**Comment by problame** [comment]
I think at this point it's mostly a question of consolidating the various envconst's and wiring them up to the config.
Can I motivate you for this task?

**Comment by InsanePrawn** [comment]
Maybe. It's probably better to edit in a short overview of what is left to do in the initial post, add the Good First Issue tag and leave it there for whoever gets to it first.

**Comment by problame** [comment]
For the record. Current variables

```
export ZREPL_REPLICATION_EXPERIMENTAL_REPLICATION_CONCURRENCY=20
export ZREPL_ENDPOINT_MAX_CONCURRENT_SEND=20
export ZREPL_ENDPOINT_MAX_CONCURRENT_RECV=20
export ZREPL_REPLICATION_MAX_CONCURRENT_SIZE_ESTIMATE=20
```
